redis/server.hh

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (C) 2019 pengjian.uestc @ gmail.com
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "auth/service.hh"
+#include "service/storage_proxy.hh"
+#include "service_permit.hh"
+#include "redis/query_processor.hh"
+#include "redis/request.hh"
+#include "cql3/values.hh"
+#include "auth/authenticator.hh"
+#include "timeout_config.hh"
+#include <memory>
+#include <seastar/net/tls.hh>
+#include <seastar/core/semaphore.hh>
+#include "seastar/core/distributed.hh"
+#include "seastar/core/reactor.hh"
+#include <seastar/core/metrics_registration.hh>
+#include "utils/fragmented_temporary_buffer.hh"
+#include "redis/protocol_parser.hh"
+#include "redis/request.hh"
+#include "redis/reply.hh"
+#include "redis/options.hh"
+#include  "service/client_state.hh"
+
+db::consistency_level make_consistency_level(const sstring&);
+
+class redis_exception;
+
+namespace redis_transport {
+
+struct redis_server_config {
+    ::timeout_config _timeout_config;
+    size_t _max_request_size;
+    db::consistency_level _read_consistency_level;
+    db::consistency_level _write_consistency_level;
+};
+
+class redis_server {
+    std::vector<server_socket> _listeners;
+    distributed<service::storage_proxy>& _proxy;
+    distributed<redis::query_processor>& _query_processor;
+    redis_server_config _config;
+    size_t _max_request_size;
+    semaphore _memory_available;
+    seastar::metrics::metric_groups _metrics;
```

Comment:
I see that you define a metric_gruup object, which is good, but I don't see any metrics defined.

Adding metrics can be in a follow-up patch, what is important is that you'll have a counter (i.e. ever-growing unsigned int) for each of your commands, so it can be monitored separately.

Also, I suggest monitoring the latencies of your read and write, take and store a timestamp at the beginning of each operation and use an estimated_histogram for latencies calculations.

Reply:
Hi @amnonh , thanks for your suggestion about metrics.
Now, only some basic metrics are added.

Reply:
By the way, you can see examples of metrics similar to what you may want to add in Alternator - alternator/stats.{cc,hh}